### PR TITLE
GreenField: Add Invoice Payment methods endpoints

### PR DESCRIPTION
**Pull request by Kukks**
lets you fetch all active payment methods data + payments made

### CODE REVIEW

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 1]
```diff
@@ -840,9 +840,11 @@ public async Task InvoiceTests()
                 Assert.Single(invoices);
                 Assert.Equal(newInvoice.Id, invoices.First().Id);
 
-                //get payment request
+                //get
                 var invoice = await viewOnly.GetInvoice(user.StoreId, newInvoice.Id);
                 Assert.Equal(newInvoice.Metadata, invoice.Metadata);
+                var paymentMethods = await viewOnly.GetInvoicePaymentMethods(user.StoreId, newInvoice.Id);
```
Comment (dennisreimann) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:846]:
```suggestion
                var paymentMethods = await viewOnly.GetInvoicePaymentMethods(user.StoreId, newInvoice.Id);
                Assert.Equal(1, paymentMethods.Length);
                var paymentMethod = paymentMethods.First();
                Assert.Equal("BTC", paymentMethod.PaymentMethod);
                Assert.Equal(0, paymentMethod.Payments.Count);
```

File: BTCPayServer.Client/Models/InvoicePaymentMethodDataModel.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            [JsonProperty(ItemConverterType = typeof(NumericStringJsonConverter))]
+            public decimal Fee { get; set; }
+
+            [JsonConverter(typeof(StringEnumConverter))]
+            public PaymentStatus Status { get; set; }
+
+            public string Destination { get; set; }
+        
+        
+
```
Comment (dennisreimann) [inline BTCPayServer.Client/Models/InvoicePaymentMethodDataModel.cs:54]:
```suggestion
```

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json [hunk 3]
```diff
@@ -565,6 +682,112 @@
...
+                    "amount": {
+                        "type": "string",
+                        "format": "decimal",
+                        "description": "The invoice amount, converted to this payment method's currency"
+                    },
+                    "networkFee": {
+                        "type": "string",
+                        "format": "decimal",
+                        "description": "The added merchant fee to pay for network costs of this payment method."
+                    },
```
Comment (dennisreimann) [inline BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json:732]: Those seem to be numbers, not strings. Here's what I received:

```json
{
        "rate": 12689.925,
        "paymentMethodPaid": 0.03573623,
        "totalPaid": 0.03573623,
        "due": 0.0,
        "amount": 0.0,
        "networkFee": 0.0,
}
```

Same for the decimals in `payments`.

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json [hunk 4]
```diff
@@ -565,6 +682,112 @@
...
+                "properties": {
+                    "id": {
+                        "type": "string",
+                        "description": "A unique identifier for this payment"
+                    },
+                    "receivedDate": {
+                        "type": "string",
+                        "format": "date-time",
+                        "description": "The date the payment was recorded"
+                    },
```
Comment (dennisreimann) [inline BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json:755]: This currently returns an UNIX timestamp, e.g. `"receivedDate": 1603287277`

File: BTCPayServer.Client/Models/InvoicePaymentMethodDataModel.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        [JsonProperty(ItemConverterType = typeof(NumericStringJsonConverter))]
+        public decimal Rate { get; set; }
+
+        [JsonProperty(ItemConverterType = typeof(NumericStringJsonConverter))]
+        public decimal PaymentMethodPaid { get; set; }
+
+        [JsonProperty(ItemConverterType = typeof(NumericStringJsonConverter))]
+        public decimal TotalPaid { get; set; }
+
+        [JsonProperty(ItemConverterType = typeof(NumericStringJsonConverter))]
```
Comment (NicolasDorier) [inline BTCPayServer.Client/Models/InvoicePaymentMethodDataModel.cs:23]: As @dennisreimann said, you don't serialize right. This should be `JsonConverter` the one you use now is only for items in arrays.

Comment (Kukks) [reply BTCPayServer.Client/Models/InvoicePaymentMethodDataModel.cs:23]: fixed

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json (hunk 3 above)
Comment (Kukks) [reply BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json:732]: fixed

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json (hunk 4 above)
Comment (Kukks) [reply BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json:755]: fixed format

**Comment by NicolasDorier** [comment]
I use BTCpayServer everyday and I could not understand the enum without looking the code 

```json
                "enum": [
                    "Invalid",
                    "AwaitingConfirmation",
                    "AwaitingCompletion",
                    "Complete"
                ]
```

**Comment by Kukks** [comment]
> I use BTCpayServer everyday and I could not understand the enum without looking the code
> 
> ```json
>                 "enum": [
>                     "Invalid",
>                     "AwaitingConfirmation",
>                     "AwaitingCompletion",
>                     "Complete"
>                 ]
> ```
> 
> Let's remove differences between "Complete" and "Confirmed"
> Just use "AwaitingConfirmation/Confirmed"
> 
> I will work soon at making a PR which make the passage from Confirm to Complete always the same, so they are basically equivalent.

done

**Review by rockstardev — APPROVED** [review]
This is in merge-able state, the only two subjective decisions for Emperor @NicolasDorier:

1. `PaymentStatus` - I was proposing Invalid, Incoming, Complete... final naming up to Nicolas.
2. Decision was made to merge AwaitingConfirmation & AwaitingCompletion... where if 6 confirmations are set by store 1st confirmation would move from AwaitingConfirmation to AwaitingCompletion into single state.

Everything else looks great, all that is left are these subjective decisions.

**Comment by NicolasDorier** [comment]
Let's make cookie happy then! :p

**Pull request merged**